src/render_helpers/mod.rs

```diff
@@ -127,7 +132,75 @@ pub fn render_to_dmabuf(
         }
     }
 
-    let _sync_point = frame.finish().context("error finishing frame")?;
+    let sync_point = frame.finish().context("error finishing frame")?;
+
+    Ok(sync_point)
+}
+
+pub fn render_to_shm(
+    renderer: &mut GlesRenderer,
+    buffer: &WlBuffer,
+    region: Rectangle<i32, Physical>,
+    size: Size<i32, Physical>,
+    scale: Scale<f64>,
+    transform: Transform,
+    elements: impl Iterator<Item = impl RenderElement<GlesRenderer>>,
+) -> anyhow::Result<SyncPoint> {
+    let buffer_size = smithay::backend::renderer::buffer_dimensions(buffer).unwrap();
+    let offscreen_buffer: GlesRenderbuffer = renderer
+        .create_buffer(Fourcc::Abgr8888, buffer_size)
+        .context("error creating renderbuffer")?;
+    renderer
+        .bind(offscreen_buffer)
+        .context("error binding renderbuffer")?;
+
+    let output_rect = Rectangle::from_loc_and_size((0, 0), size);
+
+    let mut frame = renderer
+        .render(size, transform)
+        .context("error starting frame")?;
+
+    frame
+        .clear([0., 0., 0., 0.], &[output_rect])
+        .context("error clearing")?;
+
+    for element in elements {
+        let src = element.src();
+        let dst = element.geometry(scale);
+
+        if let Some(mut damage) = output_rect.intersection(dst) {
+            damage.loc -= dst.loc;
+            element
+                .draw(&mut frame, src, dst, &[damage])
+                .context("error drawing element")?;
+        }
+    }
+
+    let sync_point = frame.finish().context("error finishing frame")?;
+
+    shm::with_buffer_contents_mut(buffer, |shm_buffer, shm_len, buffer_data| {
+        // The buffer prefers pixels in little endian ...
+        if !(buffer_data.format == wl_shm::Format::Argb8888
+            && buffer_data.stride == region.size.w * 4
+            && buffer_data.height == region.size.h
```

Comment:
Can this part be validated ahead of time, at the point when the client gives us the buffer?

Reply:
I did this in https://github.com/YaLTeR/niri/pull/243/commits/005a4626d54f8db3417a2a1bba161614e8a053f3 - but ultimately i feel that it is just worse? Because, now, this function is unsafe. It moves the concern of the buffer into the caller, and for what benefit? Performance? I think this function is much nicer when it validates the buffer internally.

I suppose we could verify it at both points (i.e. remove `debug_assert!()` macro around the condition), then the function still has the same safety guarantees but we exit early in screencopy process if the buffer is malformed.

Reply:
> I suppose we could verify it at both points (i.e. remove debug_assert!() macro around the condition), then the function still has the same safety guarantees but we exit early in screencopy process if the buffer is malformed.

Yes, I think that's a good approach.